File: doubly_linked_list/doubly_linked_list.go
package main

import "fmt"

type Node struct {
    Value int
    Next  *Node
    Prev  *Node
}

type DoublyLinkedList struct {
    Head   *Node
    Tail   *Node
    Length int
}

func NewDoublyLinkedList(value int) *DoublyLinkedList {
    newNode := &Node{Value: value}
    return &DoublyLinkedList{
        Head:   newNode,
        Tail:   newNode,
        Length: 1,
    }
}

func (dll *DoublyLinkedList) PrintList() {
    temp := dll.Head
    for temp != nil {
        fmt.Println(temp.Value)
        temp = temp.Next
    }
}

func (dll *DoublyLinkedList) Append(value int) bool {
    newNode := &Node{Value: value}
    if dll.Length == 0 {
        dll.Head = newNode
        dll.Tail = newNode
    } else {
        dll.Tail.Next = newNode
        newNode.Prev = dll.Tail
        dll.Tail = newNode
    }
    dll.Length++
    return true
}

func (dll *DoublyLinkedList) Pop() *Node {
    if dll.Length == 0 {
        return nil
    }
    temp := dll.Tail
    if dll.Length == 1 {
        dll.Head = nil
        dll.Tail = nil
    } else {
        dll.Tail = dll.Tail.Prev
        dll.Tail.Next = nil
        temp.Prev = nil
    }
    dll.Length--
    return temp
}

func (dll *DoublyLinkedList) Prepend(value int) {
    newNode := &Node{Value: value}
    if dll.Length == 0 {
        dll.Head = newNode
        dll.Tail = newNode
    } else {
        newNode.Next = dll.Head
        dll.Head.Prev = newNode
        dll.Head = newNode
    }
    dll.Length++
}

func (dll *DoublyLinkedList) PopFirst() *Node {
    if dll.Length == 0 {
        return nil
    }
    temp := dll.Head
    if dll.Length == 1 {
        dll.Head = nil
        dll.Tail = nil
    } else {
        dll.Head = dll.Head.Next
        dll.Head.Prev = nil
        temp.Next = nil
    }
    dll.Length--
    return temp
}

func (dll *DoublyLinkedList) Get(index int) *Node {
    if index < 0 || index >= dll.Length {
        return nil
    }
    var temp *Node
    if index < dll.Length/2 {
        temp = dll.Head
        for i := 0; i < index; i++ {
            temp = temp.Next
        }
    } else {
        temp = dll.Tail
        for i := dll.Length - 1; i > index; i-- {
            temp = temp.Prev
        }
    }
    return temp
}

func (dll *DoublyLinkedList) SetValue(index int, value int) bool {
    temp := dll.Get(index)
    if temp != nil {
        temp.Value = value
        return true
    }
    return false
}

func (dll *DoublyLinkedList) Insert(index int, value int) bool {
    if index < 0 || index > dll.Length {
        return false
    }
    if index == 0 {
        dll.Prepend(value)
        return true
    }
    if index == dll.Length {
        dll.Append(value)
        return true
    }
    newNode := &Node{Value: value}
    before := dll.Get(index - 1)
    after := before.Next
    newNode.Prev = before
    newNode.Next = after
    before.Next = newNode
    if after != nil {
        after.Prev = newNode
    }
    dll.Length++
    return true
}

func (dll *DoublyLinkedList) Remove(index int) *Node {
    if index < 0 || index >= dll.Length {
        return nil
    }
    if index == 0 {
        return dll.PopFirst()
    }
    if index == dll.Length-1 {
        return dll.Pop()
    }
    temp := dll.Get(index)
    temp.Prev.Next = temp.Next
    if temp.Next != nil {
        temp.Next.Prev = temp.Prev
    }
    temp.Next = nil
    temp.Prev = nil
    dll.Length--
    return temp
}

func main() {
    doubleLinkList := NewDoublyLinkedList(7)
    doubleLinkList.Append(5)
    doubleLinkList.Append(6)
    doubleLinkList.Append(4)
    doubleLinkList.Append(3)
    doubleLinkList.Pop()
    doubleLinkList.Prepend(1)
    doubleLinkList.PopFirst()
    doubleLinkList.PrintList()
    fmt.Println(doubleLinkList.Get(1))
    fmt.Println(doubleLinkList.SetValue(1, 22))
    fmt.Println(doubleLinkList.Insert(4, 33))
    fmt.Println(doubleLinkList.Remove(2))
}

File: doubly_linked_list/doubly_linked_list.py
class Node:
    def __init__(self,value):
        self.value = value
        self.next = None
        self.prev = None

class DoublyLinkedList:
    def __init__ (self ,value):
        new_node = Node(value)
        self.head = new_node
        self.tail = new_node
        self.length = 1

    def print_list(self):
        temp=self.head
        while temp is not None:
            print(temp.value)
            temp = temp.next
    
    def append(self, value):
        new_node = Node(value)
        if self.head is None:
            self.head = new_node
            self.tail = new_node
        else:
            self.tail.next = new_node
            new_node.prev = self.tail
            self.tail = new_node
        self.length +=1
        return True
    
    def pop(self):
        if self.length == 0:
            return None
        temp = self.tail
        self.tail = self.tail.prev
        self.tail.next = None
        temp.prev = None
        self.length -=1
    
    def prepend(self,value):
        new_node = Node(value)
        if self.head is None:
            self.head = new_node
            self.tail = new_node
        else:
            new_node.next = self.head
            self.head.prev = new_node
            self.head = new_node
        self.length +=1
    
    def pop_first(self):
        if self.length == 0:
            return None
        temp = self.head
        if self.length == 1:
            self.head = None
            self.tail = None
        else:
            self.head = self.head.next
            self.head.prev = None
            temp.next = None
            self.length -= 1
        return temp
    
    def get(self,index):
        if index < 0 or index >= self.length:
            return None
        temp = self.head       
        if index < self.length/2:
            for _ in range(index):
                temp= temp.next
        else:
            temp = self.tail
            for _ in range(self.length -1, index,-1):
                temp= temp.prev
        return temp
    
    def set_value(self,index,value):
         temp = self.get(index)
         if temp is not None:
             temp.value = value
             return True
         else:
            return False
         
    def insert(self,index,value):
        if index <0 or index >self.length:
            return False
        if index == 0:
            return self.prepend(value)
        if index == self.length:
            return self.append(value)
        new_node = Node(value)
        before = self.get(index -1)
        after = before.next

        new_node.prev = before
        new_node.next = after
        before.next = new_node
        after.prev = new_node    
        
        self.length += 1
        return True
    
    def remove(self,index):
        if index <0 or index >self.length:
            return False
        if index == 0:
            return self.pop()
        if index == self.length:
            return self.pop_first()
        
        temp = self.get(index)
        temp.next.prev = temp.prev
        temp.prev.next = temp.next
        temp.next = None
        temp.prev = None

        self.length -= 1
        return temp
        

double_link_list = DoublyLinkedList(7)
double_link_list.append(5)
double_link_list.append(6)
double_link_list.append(4)
double_link_list.append(3)
double_link_list.pop()
double_link_list.prepend(1)
double_link_list.pop_first()
double_link_list .print_list()
print(double_link_list.get(1))
print(double_link_list.set_value(1,22))
print(double_link_list.insert(4,33))
print(double_link_list.remove(2))